### ISSUE TITLE
change wireless keyboard

### ISSUE THREAD

**Issue by youxiaojie** [robszy/ps2usb#1]
can I modify this code to convert spi protocol to USB keyboard mouse and make a wireless connections use nrf24l01?

**Comment by robszy** [comment]
Hi

You want to change code to carry keyboard and mouse signals over wireless
spi and at receiver translate to usb ?

czw., 10 sty 2019 o 14:53 youxiaojie <notifications@github.com> napisał(a):

**Comment by youxiaojie** [comment]
certainly. I see there's sources of nrf24l01 chip to transfer data by spi, so i hope to modify your source code to use this. firstly use a transmiter (arduino eg atmega328p) to translate scancode to spi and then receiver (arduino /teensy atmega32u4) to do the following this eventually translate ps/2 keyboard's scancode to hid code. do you think is it feasible?

I want to experiment by conncet two arduino by spi wire  first, if ok add wireless code. is it feasible?
thank you.
and another question, is your converter single direction or bi-direction? could the keyboard receive the command from host?
thank you again.

**Comment by youxiaojie** [comment]
Aim of this project is to make _usable (not like translators on the market :) ) translator_ for keyboard and mouse which uses ps2 interface to connect them to new PC's which have only usb ports.

firstly what does this sentence mean? what is the different between yours and market's converter/ translator?
and secondly it seem like your converter realize the usb protocol by programming? am i right? I see your code is more complex than others eg tmk which used atmega32u4 and pjrc's ps/2 routine usually.
V-USB Support

With V-USB you can use this converter on ATmega(168/328) but it doesn't support NKRO at this time.

Circuit:
```
                +---+   +---------------+
USB            GND  |   |   ATmega168   |
===                 C3  |               |
5V <-------+--------+---|Vcc,AVCC       |        PS/2
           R1           |               |        ====
D- <----+--+-----R2-----|INT1        RXD|------->DATA
D+ <----|---+----R3-----|INT0        XCK|------->CLOCK
        Z1  Z2          |               |      ->5V
GND<----+---+--+--+-----|GND            |      ->GND
               |  |     |               |
               |  C2-+--|XTAL1          |
               |     X1 |               |
               +--C3-+--|XTAL2          |
                        +---------------+
R1:     1.5K Ohm
R2,R3:  68 Ohm
Z1,Z2:  Zenner 3.6V
C1,C2:  22pF
C3:     0.1uF
X1:     Crystal 20MHz(16MHz/12MHz)
```

is it your way to deal with ps/2 signal by atmega 328?

**Comment by robszy** [comment]
Sorry for late response

pt., 11 sty 2019 o 18:41 youxiaojie <notifications@github.com> napisał(a):

> certainly. I see there's sources of nrf24l01 chip to transfer data by spi,
> so i hope to modify your source code to use this. firstly use a transmiter
> (arduino eg atmega328p) to translate scancode to spi and then receiver
> (arduino /teensy atmega32u4) to do the following this eventually translate
> ps/2 keyboard's scancode to hid code. do you think is it feasible?
>
Its ok but i would transmit only hid code to receiver

> I want to experiment by conncet two arduino by spi wire first, if ok add
> wireless code. is it feasible?
> thank you.
> and another question, is your converter single direction or bi-direction?
> could the keyboard receive the command from host?
>
USB must be bidirectional connection and it actualy receives commands which
i use to switch leds

**Comment by robszy** [comment]
pt., 11 sty 2019 o 18:55 youxiaojie <notifications@github.com> napisał(a):

> Aim of this project is to make *usable (not like translators on the
> market :) ) translator* for keyboard and mouse which uses ps2 interface
> to connect them to new PC's which have only usb ports.
>
> firstly what does this sentence mean? what is the different between yours
> and market's converter/ translator?
>
Market translator have poor performance. They have all sorts of bugs like
repeat last stroked key forever switching of pressed key and others

> and secondly it seem like your converter realize the usb protocol by
> programming? am i right? I see your code is more complex than others eg tmk
> which used atmega32u4 usually and pjrc's ps/2 routine.
> V-USB Support
>
Yes i use Vusb and it is not easy to understand in usb asm routines but i
think you should just copy it in receiver side which will be just sending
received hid codes

> With V-USB you can use this converter on ATmega(168/328) but it doesn't
> support NKRO at this time.
>
NRKO is keyboard feature not translator

> Circuit:
>
>                 +---+   +---------------+
> USB            GND  |   |   ATmega168   |
> ===                 C3  |               |
> 5V <-------+--------+---|Vcc,AVCC       |        PS/2
>            R1           |               |        ====
> D- <----+--+-----R2-----|INT1        RXD|------->DATA
> D+ <----|---+----R3-----|INT0        XCK|------->CLOCK
>         Z1  Z2          |               |      ->5V
> GND<----+---+--+--+-----|GND            |      ->GND
>                |  |     |               |
>                |  C2-+--|XTAL1          |
>                |     X1 |               |
>                +--C3-+--|XTAL2          |
>                         +---------------+
> R1:     1.5K Ohm
> R2,R3:  68 Ohm
> Z1,Z2:  Zenner 3.6V
> C1,C2:  22pF
> C3:     0.1uF
> X1:     Crystal 20MHz(16MHz/12MHz)
>
> is it your way to deal with ps/2 signal by atmega 328?
>
I use one of int to handle keyboard and uart for mouse (i'm not sure if
INT1 is connected to D- in my translator)

**Comment by youxiaojie** [comment]
thanks after I read source code of lots of PS/2 USB converter I decide transfer USB HID. among these projects 1yours 2 https://github.com/nsd20463/ps2_kbd_to_usb_adapter/issues/3
 3 tmk is preferred, others are based in arduino majorly, thanks for advice very much. thanks for your patience to instruct me.

**Comment by youxiaojie** [comment]
what is this means？
extern const uint8_t keymaps[][MATRIX_ROWS][MATRIX_COLS];
is it a pointer？for a simplified example int p[][3][4] means？

**Comment by robszy** [comment]
It seams but you shuld ask tmk ;)

pon., 28 sty 2019 o 19:15 youxiaojie <notifications@github.com> napisał(a):

**Comment by youxiaojie** [comment]
my dear friend and teacher,
build with atmel studio 7,some errors in keyboard.c,
some is 
 Severity	Code	Description	Project	File	Line
Error		'odDebugInit' undeclared (first use in this function)	ps2usb	K:\ps2usb-master\ps2usb\src\keyboard.c	159
Severity	Code	Description	Project	File	Line
Error		macro "odDebugInit" passed 1 arguments, but takes just 0	ps2usb	K:\ps2usb-master\ps2usb\src\keyboard.c	159
Severity	Code	Description	Project	File	Line
Error		implicit declaration of function 'waitForEnd' [-Werror=implicit-function-declaration]	ps2usb	K:\ps2usb-master\ps2usb\src\keyboard.c	178

why the new errors appear?

Severity	Code	Description	Project	File	Line
Error		implicit declaration of function 'cli' [-Werror=implicit-function-declaration]	ps2usb	K:\ps2usb-master\ps2usb\src\keyboard.c	117
Error		implicit declaration of function '_delay_ms' [-Werror=implicit-function-declaration]	ps2usb	K:\ps2usb-master\ps2usb\src\keyboard.c	127
Error		implicit declaration of function 'sei' [-Werror=implicit-function-declaration]	ps2usb	K:\ps2usb-master\ps2usb\src\keyboard.c	134
this three errors I correct it by adding #include delay.h and interrupt.h.
and the left, I have no way and have to ask you

the last question, can you remember what is the pin connexion of two ps/2 and usb?
thanks again

**Comment by robszy** [comment]
Try release build and my studio version is 6.0.1996 sp2

wt., 29 sty 2019 o 20:25 youxiaojie <notifications@github.com> napisał(a):
